// unexpected-eventemitter/dist/unexpected-eventemitter.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.unexpectedEventEmitter = f()}})(function(){var define,module,exports;return (function(){function r(e,n,t){function o(i,f){if(!n[i]){if(!e[i]){var c="function"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error("Cannot find module '"+i+"'");throw a.code="MODULE_NOT_FOUND",a}var p=n[i]={exports:{}};e[i][0].call(p.exports,function(r){var n=e[i][1][r];return o(n||r)},p,p.exports,r,e,n,t)}return n[i].exports}for(var u="function"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o}return r})()({1:[function(require,module,exports){
module.exports={
  "version": "2.2.0"
}
},{}],2:[function(require,module,exports){
'use strict';

var ref = require('../package.json');
var version = ref.version;

module.exports = {
  name: 'unexpected-eventemitter',
  version: version,
  installInto: function installInto(expect) {
    expect.addType({
      base: 'object',
      name: 'EventEmitter',
      identify: function (obj) { return obj !== null &&
        typeof obj === 'object' &&
        typeof obj.emit === 'function' &&
        typeof obj.once === 'function' &&
        typeof obj.on === 'function'; },
    });

    expect
      .addAssertion(
        '<function> [not] to emit from <EventEmitter> <string> <any*>',
        function (expect, subject, ee, eventName) {
          var values = [], len = arguments.length - 4;
          while ( len-- > 0 ) values[ len ] = arguments[ len + 4 ];

          var emitted = false;
          var emittedValues;
          // expect.errorMode = 'bubble';

          var onEvent = function () {
            var values = [], len = arguments.length;
            while ( len-- ) values[ len ] = arguments[ len ];

            emitted = true;
            emittedValues = values;
          };

          ee.once(eventName, onEvent);

          return expect
            .promise(function () {
              expect.errorMode = 'bubble';
              return expect(subject, 'not to error');
            })
            .then(function () {
              expect.errorMode = 'default';
              expect(emitted, '[not] to be true');
              expect.errorMode = 'nested';
              values.forEach(function (value, idx) {
                expect(emittedValues[idx], '[not] to satisfy', value);
              });
            })
            .finally(function () {
              ee.removeListener(eventName, onEvent);
            });
        }
      )
      .addAssertion(
        '<function> to emit with error from <Error> <EventEmitter> <string> <any*>',
        function (expect, subject, error, ee, eventName) {
          var values = [], len = arguments.length - 5;
          while ( len-- > 0 ) values[ len ] = arguments[ len + 5 ];

          var emitted = false;
          var emittedValues;

          var onEvent = function () {
            var values = [], len = arguments.length;
            while ( len-- ) values[ len ] = arguments[ len ];

            emitted = true;
            emittedValues = values;
          };

          ee.once(eventName, onEvent);

          return expect
            .promise(function () {
              expect.errorMode = 'bubble';
              return expect(subject, 'to error with', error);
            })
            .then(function () {
              expect.errorMode = 'default';
              expect(emitted, 'to be true');
              expect.errorMode = 'nested';
              values.forEach(function (value, idx) {
                expect(emittedValues[idx], 'to satisfy', value);
              });
            })
            .finally(function () {
              ee.removeListener(eventName, onEvent);
            });
        }
      );
  },
};

},{"../package.json":1}]},{},[2])(2)
});
